modin/core/dataframe/algebra/binary.py

```diff
@@ -413,3 +413,47 @@ def caller(
                 )
 
         return caller
+
+    @classmethod
+    def apply(
+        cls, left, right, func, axis=0, func_args=None, func_kwargs=None, **kwargs
+    ):
+        r"""
+        Apply a binary function row-wise or column-wise.
+
+        Parameters
+        ----------
+        left : modin.pandas.DataFrame or modin.pandas.Series
+            Left operand.
+        right : modin.pandas.DataFrame or modin.pandas.Series
+            Right operand.
+        func : callable(pandas.DataFrame, pandas.DataFrame, \*args, axis, \*\*kwargs) -> pandas.DataFrame
+            A binary function to apply `left` and `right`.
+        axis : int, default: 0
+            Whether to apply the function across rows (``axis=0``) or across columns (``axis=1``).
+        func_args : tuple, optional
+            Positional arguments to pass to the funcs.
+        func_kwargs : dict, optional
+            Keyword arguments to pass to the funcs.
+        **kwargs : dict
+            Additional arguments to pass to the ``cls.register()``.
+
+        Returns
+        -------
+        The same type as `df`.
+        """
+        from modin.pandas import Series
```

Comment:
I don't really like that inner layers depend on higher layers. I do not see any benefit of introducing these changes rather than it will simplify registriation of a UDF for users, which doesn't happen very often from my point of view. I would like to bring more attention to these changes to decide whether we want these changes to be merged or not.

cc @modin-project/modin-core

Reply:
maybe we can make the `.apply()` method a separate function and place it somewhere at `modin.pandas.utils`, the function would look something like:
```python
# modin/pandas/utils.py
def apply_operator(df : modin.pandas.DataFrame, operator_cls : type[Operator], func, *args, **kwargs):
    res_qc = operator_cls.register(func)(df._query_compiler, *args, **kwargs)
    return type(df)(query_compiler=res_qc)

# the usage then would be:

from modin.pandas.utils import apply_operator
from modin.core.dataframe.algebra import Reduce

res_df = apply_operator(df, Reduce, func=reduce_func, axis=1)
```

One of the problem I see here is that managing operators-dependent behavior can be quite a pain here, as we can no longer use OOP mechanisms to align with operator-specific logic using inheritance and overriding: 

```python
# modin/pandas/utils.py

def _apply_reduce_op(...):
   ...

def _apply_groupby_op(...):
   ...
...

_operators_dict = {
    Reduce: _apply_reduce_op,
    GroupbyReduce: _apply_groupby_op,
    ...
}

def apply_operator(df : modin.pandas.DataFrame, operator_cls : type[Operator], func, *args, **kwargs):
    return _operators_dict[operator_cls](df, func, *args, **kwargs)
```

Do you have any other suggestion on how to improve this approach (or maybe you have another approach in mind)?

Reply:
@dchigarev  to me the [current approach](https://modin.readthedocs.io/en/stable/flow/modin/core/dataframe/algebra.html#how-to-register-your-own-function) for using the operators doesn't look very verbose to me. will this PR make it much easier to use the operators anywhere?

Reply:
> will this PR make it much easier to use the operators anywhere

The PR should make the usage of operators much easier for end-users of modin who would like to define their own distributed functions using modin's operators.

While optimizing customer workloads for modin we sometimes see places that would perform much better if rewritten from pandas API using modin's operators, however [the present API the operators provide](https://modin.readthedocs.io/en/stable/flow/modin/core/dataframe/algebra.html#how-to-register-your-own-function) causes a lot of complex code written around that customers struggle to understand. That inspired us to create a simple method/function that makes operator's usage [as simple as calling one single function](https://modin--6302.org.readthedocs.build/en/6302/flow/modin/core/dataframe/algebra.html#how-to-use-udfs-with-these-operators).

Reply:
> I don't really like that inner layers depend on upper layers.

So I made two versions of how we can eliminate this dependency.

1. Avoid importing objects from `modin.pandas...` to the algebra level, but still allow passing such objects to the `Operator.apply()`. This way we're getting rid of the 'import dependency' on the higher level, meaning that we can easily detach the algebra layer if needed without worrying that it would require to bear stuff from the higher levels for the algebra to work correctly.

   I've made the changes to align with this approach and pushed it to the branch from this PR.

2. Also avoid passing dataframe objects to the `Operator.apply()` and rework this method to accept query compilers only. Then add a helper function somewhere to the dataframe level that would take series/dataframes, extract their QCs, and pass to them the `Operator.apply()`.

    I've implemented this approach in [a separate branch](https://github.com/dchigarev/modin/commit/1a8f23e37a2441d1efc470ed7bde2f3526d805e1). There, users have a function at [`modin.pandas.utils.apply_operator`](https://github.com/dchigarev/modin/blob/1a8f23e37a2441d1efc470ed7bde2f3526d805e1/modin/pandas/utils.py#L375) with the following signature:

    ```python
   def apply_operator(operator_cls, *args, **kwargs):
        pass
    ...
    # use case example
    from modin.pandas.utils import apply_operator
    from modin.core.dataframe.algebra import Reduce

    series_obj = apply_operator(Reduce, df, reduce_func, axis=1)
   ```

    I don't really like this approach as `apply_operator()` doesn't provide meaningful signature and requires referring to the `Operator.apply()` for the list of allowed parameters.

Reply:
@mvashishtha @YarShev thoughts?

Reply:
@dchigarev I wonder if you could provide an example of user-defined Modin operator (ideally a real case, even if simplified and anonymized)

Reply:
> I wonder if you could provide an example of user-defined Modin operator

@vnlitvinov 

What we usually use the user-defined operators for is to emulate lazy execution for those types of transformations that can't be written into a one or two pandas API calls (usually such transformations are performed in a for-loop).

As an example, consider that we have a dataframe with multiple string columns, and then we want to combine those columns into a single one using the specified separator. Surprisingly, but the fastest way to do so using vanilla pandas is simply writing a for-loop:

```
combining multiple string columns using different approaches in pandas:

for-loop: 17.533942513167858
df.apply(join): 21.597900850698352
df.str.cat(): 38.55254930164665
```

<details><summary>pandas code</summary>

```python
import pandas as pd
import numpy as np
from timeit import default_timer as timer

NCOLS = 16
NROWS = 5_000_000

df = pd.DataFrame({f"col{i}": [f"col{i}-{j}" for j in range(NROWS)] for i in range(NCOLS)})

t1 = timer()
res = df.iloc[:, 0]
for col in df.columns[1:]:
    res += "_" + df[col]
print(f"for-loop: {timer() - t1}")

t1 = timer()
res = df.apply(lambda row: "_".join(row), axis=1)
print(f"df.apply(join): {timer() - t1}")

t1 = timer()
res = df.iloc[:, 0].str.cat(df.iloc[:, 1:], sep="_")
print(f"df.str.cat(): {timer() - t1}")
```

</details>

Then when adapting this code to modin, it appears that the for-loop approach works very slow due to a lot of kernels being submitted to Ray and so causing it to overwhelm (each iteration of the for-loop will result into 3 separate kernels: 1. `df[col]`; 2. `"_" + df[col]`; 3. `res +=`). And then it appears, that the most performant approach is to submit this for-loop as a single kernel using the Reduction operator:

```
combining multiple string columns using different approaches in modin:

reduction operator: 2.6848975336179137
batch pipeline API: 2.945119895040989
for-loop: 36.92861177679151
df.apply(join): 8.54124379903078
df.str.cat(): 43.84469765238464
```

<details><summary>modin code</summary>

```python
import modin.pandas as pd
import modin.config as cfg
import numpy as np
from timeit import default_timer as timer

cfg.BenchmarkMode.put(True)
# start all the workers
pd.DataFrame([np.arange(cfg.MinPartitionSize.get()) for _ in range(cfg.NPartitions.get() ** 2)]).to_numpy()

NCOLS = 16
NROWS = 5_000_000

df = pd.DataFrame({f"col{i}": [f"col{i}-{j}" for j in range(NROWS)] for i in range(NCOLS)})

from modin.core.dataframe.algebra import Reduce

def reduction(df):
    res = df.iloc[:, 0]
    for col in df.columns[1:]:
        res += "_" + df[col]
    return res

t1 = timer()
res = Reduce.apply(df, reduction, axis=1)
print(f"reduction operator: {timer() - t1}")

from modin.experimental.batch import PandasQueryPipeline

t1 = timer()
pipeline = PandasQueryPipeline(df)
pipeline.add_query(reduction, is_output=True)
res = pipeline.compute_batch()
print(f"batch pipeline API: {timer() - t1}")

t1 = timer()
res = df.iloc[:, 0]
for col in df.columns[1:]:
    res += "_" + df[col]
print(f"for-loop: {timer() - t1}")

t1 = timer()
res = df.apply(lambda row: "_".join(row), axis=1)
print(f"df.apply(join): {timer() - t1}")

t1 = timer()
res = df.iloc[:, 0].str.cat(df.iloc[:, 1:], sep="_")
print(f"df.str.cat(): {timer() - t1}")
```

</details>

(as I was writing this comment, I found out about the [batch API in modin](https://modin.readthedocs.io/en/stable/usage_guide/advanced_usage/batch.html) that is supposed to serve exactly the same purpose of "emulating" the lazy execution. However, it seems that it doesn't provide a way to specify the scheme on how the kernels actually should be submitted (map, row-wise, column-wise, ...) and also have some slight overhead when comparing with the pure user-defined operator's approach)